conans/client/rest_client_local_recipe_index.py

```diff
@@ -76,7 +77,12 @@ def get_recipe(self, ref, dest_folder):
         return self._copy_files(export_folder, dest_folder)
 
     def get_recipe_sources(self, ref, dest_folder):
-        export_sources = self._app.cache.recipe_layout(ref).export_sources()
+        try:
```

Comment:
Is this not necessary in `get_recipe()` above?

Reply:
Hi @cynix 

I am trying to understand the root cause and the necessary changes to fix it.

So far, I have only been able to make it fail for the ``get_recipe_sources()`` case, when Conan iterates the remotes looking for the sources of a given (already existing) recipe. The above ``get_recipe()`` seems to be sufficiently protected by previous checks, but the sources one wasn't. As you can see, the test keeps working (we do quite test-focused development, if there is something else that could be failing in the ``get_recipe()`` it would be necessary another test that makes it fail).

According to the above, I have removed the extra calls (to avoid extra cost, those will do extra queries to the DB) and just converted the exception to make it work.

Have you been able to test the changes against your full code? Does it work? If not, maybe you can try to provide a new test that makes it fail again?

Reply:
You're right, I tested this in my full setup and it does address the issue just fine, as far as I can tell.

But if you have some time, I'd appreciate if you could elaborate on why `get_recipe` was already "protected" but `get_recipe_sources` was not, and whether this is a symptom of a bigger issue. It seems the calling code _should_ have an idea of which remote a recipe comes from (after all, only one remote can possibly have the latest revision, right?). Once we've determined which remote a recipe comes from, should we then only ask that remote for sources? Does it make sense to still ask all remotes? When I was originally looking into this issue, an alternative I considered was to change [the caller](https://github.com/conan-io/conan/blob/7f3f09945f8972f1fb37da38fa325dab11a52390/conans/client/installer.py#L194) to pass in a single remote (e.g. `node.remote`) instead of the full list of enabled remotes, but I wasn't sure if that's intended or not.

Reply:
Sure. In the ``DepsGraphBuilder`` and ``ConanProxy`` class, the logic is more or less:
- Do first a `` ref = self._remote_manager.get_latest_recipe_reference(reference, remote)`` to query about the existing ``reference`` in the remote.
- If there is such a ``reference``, then do a ``remote_manager.get_recipe()`` to get it

While the ``remote_manager.get_recipe_sources()`` is called in a different place (only when something needs to build); protected with:
```python
def _try_get_sources(ref, remote_manager, recipe_layout, remote):
    try:
        remote_manager.get_recipe_sources(ref, recipe_layout, remote)
    except NotFoundException:
        return
```
And this ``_try_get_sources`` is iterated for every remote, without checking first if the ``ref`` exists in the remote. This is the operation that was not raising a ``NotFoundException`` and breaking that case.

Reply:
To further clarify, the first resolution of the ``ref`` in the graph resolution is often giving a ``reference`` without recipe-revision, and returning the full reference with recipe revision in ``ref = self._remote_manager.get_latest_recipe_reference(reference, remote)``. The ``get_recipe_sources()`` already works with a fully defined reference with recipe-revision, so no need to query for the latest one, just ask the server "give me the sources of this specific recipe revision, because I need them", and it can directly manage the ``NotFoundException``, as it is more efficient than doing 2 calls, one to check the existence, then another to retrieve the sources.

Reply:
Please let me know if this clarifies a bit, and we can proceed to merge it. Thanks!

Reply:
Exactly, the recipe-revision guarantees that you are getting the same sources irrespective of the remote, so the first one containing it will be good.

> That makes sense, but since resolving the full ref means you now know which remote it came from, isn't it still more efficient to just go directly to that remote for the sources rather than iterating through the full list again?

Yes, but Conan doesn't keep that memory/state. It did in Conan 1, but we removed it as messy (issues when remotes are removed, disabled, etc). So this is a simple enough, and retrieving the sources is almost always for building, which is very expensive, so a couple of extra api calls is neglegible.

>  Is there ever a case where a remote may provide the full ref but is unable to provide the sources?

No, that would be a corrupted package recipe.